### PR TITLE
[lldb] Invert dependency between SwiftLanguageRuntime and ObjectFile plugins

### PR DESCRIPTION
**Pull request by bulbazord**
The motivation behind this change is that ObjectFile and its derived
classes don't need to know about swift directly. SwiftLanguageRuntime
on the other hand has direct knowledge of swift so it should be able to
make the decision it needs to with the ObjectFile it has.

cc @JDevlieghere @adrian-prantl

If you'd like this on other branches just tell me which and I will open PRs for those too.

### CODE REVIEW

**Comment by adrian-prantl** [comment]
Does the llvm::ObjectFile know about Swift reflection sections? E.g., for otool & friends?

**Comment by bulbazord** [comment]
I don't think so? I see no mentions of `swift::ReflectionSectionKind` nor `swift/ABI/ObjectFile.h` included anywhere in LLVM.

**Comment by adrian-prantl** [comment]
What I'm trying to figure is — is there a good reason to instead upstream knowledge about these sections, for example to expose them in otool or related utilities? I wouldn't want to build up a second layer of object file format knowledge in LLDB if we could instead upstream it, and upstreaming it would make sense even in an llvm-project monorepo that doesn't actually include a Swift compiler. @JDevlieghere what do you think?

**Comment by JDevlieghere** [comment]
LLVM's ObjectFile doesn't seem to know about language specific sections, other than that `swift_ast` is a debug section for Mach-O. In Swift this is part of the ABI and as such I think it makes sense to have that be part of Swift rather than LLVM, and even if we upstreamed it, we probably wouldn't want to have that live strictly in LLVM. 

Ultimately this is just asking Swift for a section name and using that to get the actual section form the object file. Given that Swift already knows about ObjectFile, I think that it might make sense for this to live in Swift, but I don't know where exactly, neither ABI nor Reflection seem like good candidates. 

TL;DR I'd like Swift to have a `llvm::ObjectFile -> swift::SwiftObjectFileFormat` helper in Swift and then I think it makes sense for the rest to live in LLDB.

**Comment by bulbazord** [comment]
(Just remembered I submitted this months ago and forgot about it. Sorry!)

> TL;DR I'd like Swift to have a llvm::ObjectFile -> swift::SwiftObjectFileFormat helper in Swift and then I think it makes sense for the rest to live in LLDB.

This would make sense I think. Is there a way to go from an `lldb_private::ObjectFile` to an `llvm::ObjectFile`? I've spent a few minutes looking around but there are no references to llvm's ObjectFile class in lldb at all. As far as I'm aware, we strictly have only lldb. I could implement the `llvm::ObjectFile -> swift::SwiftObjectFileFormat` helper but there would also need to be a way for lldb to go from `lldb_private::ObjectFile -> llvm::ObjectFile`. I'm not sure that's worth the complexity. Unless I'm missing something?
